core/contract.go

```diff
@@ -106,73 +91,79 @@ func (c *Contract) Nonce() (nonce *felt.Felt, err error) {
 
 // UpdateNonce updates the nonce value in the database.
 func (c *Contract) UpdateNonce(nonce *felt.Felt) error {
-	if c.storage == nil {
-		return ErrContractNotDeployed
-	}
-
 	nonceKey := db.ContractNonce.Key(c.Address.Marshal())
 	return c.txn.Set(nonceKey, nonce.Marshal())
 }
 
-// ClassHash returns hash of the class this contract instantiates.
-func (c *Contract) ClassHash() (classHash *felt.Felt, err error) {
-	if c.storage == nil {
-		return nil, ErrContractNotDeployed
-	}
-
-	key := db.ContractClassHash.Key(c.Address.Marshal())
-	err = c.txn.Get(key, func(val []byte) error {
+// ClassHash returns hash of the class that the contract at the given address instantiates.
+func classHash(addr *felt.Felt, txn db.Transaction) (classHash *felt.Felt, err error) {
+	key := db.ContractClassHash.Key(addr.Marshal())
+	err = txn.Get(key, func(val []byte) error {
 		classHash = new(felt.Felt)
 		classHash.SetBytes(val)
 		return nil
 	})
 	return
 }
 
-// Root returns the root of the contract storage. If the Contract has not being deployed an error would be returned.
-// For a deployed contract with empty storage zero felt would be returned.
-func (c *Contract) Root() (*felt.Felt, error) {
-	if c.storage == nil {
-		return nil, ErrContractNotDeployed
-	}
-	return c.storage.Root()
+// ClassHash returns hash of the class that this contract instantiates.
+func (c *Contract) ClassHash() (*felt.Felt, error) {
+	return classHash(c.Address, c.txn)
 }
 
-func (c *Contract) Get(key *felt.Felt) (*felt.Felt, error) {
-	if c.storage == nil {
-		return nil, ErrContractNotDeployed
+// storage returns the [core.Trie] that represents the
+// storage of the contract.
+func storage(addr *felt.Felt, txn db.Transaction) (*trie.Trie, error) {
+	addrBytes := addr.Marshal()
+	var contractRootKey *bitset.BitSet
+
+	if err := txn.Get(db.ContractRootKey.Key(addrBytes), func(val []byte) error {
+		contractRootKey = new(bitset.BitSet)
+		return contractRootKey.UnmarshalBinary(val)
+	}); err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		// Don't continue normal operation with arbitrary
+		// database error.
+		return nil, err
 	}
+	trieTxn := NewTransactionStorage(txn, db.ContractStorage.Key(addrBytes))
+	return trie.NewTrie(trieTxn, contractStorageTrieHeight, contractRootKey), nil
+}
 
-	return c.storage.Get(key)
+// Root returns the root of the contract storage.
+func (c *Contract) Root() (*felt.Felt, error) {
+	if cStorage, err := storage(c.Address, c.txn); err != nil {
+		return nil, err
+	} else {
+		return cStorage.Root()
+	}
```

Comment:
@omerfirmak, it is idiomatic to do the following:
```go
func (c *Contract) Root() (*felt.Felt, error) {
	if cStorage, err := storage(c.Address, c.txn); err != nil {
		return nil, err
	} 
	return cStorage.Root()
}
```

Also when all the `if`, `else if` and `else` statements are returning there is no need to have `if else/ else` statements. We should prefer the following:

```go
classHashKey := db.ContractClassHash.Key(addr.Marshal())
err := txn.Set(classHashKey, classHash.Marshal())
if err != nil {
	return nil, err
}

c, err := NewContract(addr, txn)
if err != nil {
	return nil, err
}

err = c.UpdateNonce(&felt.Zero)
if err != nil {
	return nil, err
}

return c, nil
```

Over: 

```go
classHashKey := db.ContractClassHash.Key(addr.Marshal())
if err := txn.Set(classHashKey, classHash.Marshal()); err != nil {
	return nil, err
} else if c, err := NewContract(addr, txn); err != nil {
	return nil, err
} else if err = c.UpdateNonce(&felt.Zero); err != nil {
	return nil, err
} else {
	return c, nil
}
```
Please refer to Effective of [here](https://go.dev/doc/effective_go#if) and this [article](https://www.calhoun.io/one-liner-if-statements-with-errors/).

Reply:
Got it, thanks!